### PR TITLE
fix(windows): add null pointer checks imsample keyboard

### PR DESCRIPTION
**Pull request by rc-swag**
tl;dr  add null ptr checks to pointers before access.

The tester discovered crashed when loading the imsample keyboard after installing it.    I went through and added checks throughout the code. I built in and attached to the PR the tester was verifying. I am just now checking in this branch.

Note: I made the coding style the current Keyman style. Normally on an old project like this, I would be happy to make the new code match the current code. If that is what you prefer please comment and I will change it.

### CODE REVIEW

**Comment by rc-swag** [comment]
@keymanapp-test-bot skip

**Comment by mcdurdin** [comment]
> The tester discovered crashed when loading the imsample keyboard after installing it.
>
> keymanapp-test-bot skip

Hmm, given this was discovered in testing, shouldn't we be testing the fix?

File: windows/src/developer/samples/imsample/IMSample.txt [hunk 1]
```diff
@@ -1,6 +1,8 @@
 [Config]
 Font=Arial Unicode MS
-Size=20
+Size=15
+gridx=30
+gridy=30
```
Comment (mcdurdin) [inline windows/src/developer/samples/imsample/IMSample.txt:5]: Are these changes related to the fix?

File: windows/src/developer/samples/imsample/imsample.cpp [hunk 2]
```diff
@@ -426,10 +432,19 @@ LRESULT CALLBACK IMSampleChildWndProc(HWND hwndChild, UINT msg, WPARAM wParam, L
...
+    if (!curkbd) {
+       return 0;
+    }
+    if (!curkbd->hFont) {
+       return 0;
+    }
+    if (!curkbd->groups) {
+      return 0;
+    }
 		if(!PrepIM()) return DefWindowProc(hwndChild, msg, wParam, lParam);
```
Comment (mcdurdin) [inline windows/src/developer/samples/imsample/imsample.cpp:444]:
```suggestion
    if (!curkbd || !curkbd->hFont || !curkbd->groups || !PrepIM()) {
      return DefWindowProc(hwndChild, msg, wParam, lParam);
    }
```

You need to call `BeginPaint`/`EndPaint` on a `WM_PAINT`, or else `return DefWindowProc()` in order to validate the paint region, otherwise `WM_PAINT` will keep getting called.

File: windows/src/developer/samples/imsample/imsample.cpp [hunk 3]
```diff
@@ -566,6 +587,9 @@ extern "C" BOOL __declspec(dllexport) WINAPI DFText(HWND hwndFocus, WORD KeyStro
 {
 	WCHAR buf[32];
 	if(!curkbd) return FALSE;
+  if (!curkbd->groups) {
+    return FALSE;
+  }
```
Comment (mcdurdin) [inline windows/src/developer/samples/imsample/imsample.cpp:592]:
```suggestion
	if (!curkbd || !curkbd->groups) {
    return FALSE;
  }
```

Suggest doing this throughout your changes -- no need to split these failure conditions up.

File: windows/src/developer/samples/imsample/imsample.cpp [hunk 4]
```diff
@@ -509,7 +524,10 @@ LRESULT CALLBACK IMSampleChildWndProc(HWND hwndChild, UINT msg, WPARAM wParam, L
 
 BOOL FindRule(group *g, rule **rp, WCHAR KeyChar)
 {
-	WCHAR buf[48];
+	if(!g){
+    return FALSE;
+  }
```
Comment (mcdurdin) [inline windows/src/developer/samples/imsample/imsample.cpp:529]: When is `FindRule` being called with a `nullptr` `g`? That seems like a bug there. And if we're going to check input parameters, we should check `rp` as well:

```suggestion
	if (!g || !rp) {
    return FALSE;
  }
```

**Review by mcdurdin — CHANGES_REQUESTED** [review]

**Comment by rc-swag** [comment]
> > The tester discovered crashed when loading the imsample keyboard after installing it.
> > keymanapp-test-bot skip
> 
> Hmm, given this was discovered in testing, shouldn't we be testing the fix?

Happy to. As I said in this description " I built in and attached to the PR the tester was verifying. I am just now checking in this branch."
In other words the fixes were verified in the PR the tester was blocked on the imsample keyboard crashing, and not loading the IM window.  They then tested the fixes in this branch as I built it and attached it to the PR.

However, it was a quick fix, it still has many improvements to be made to this sample. For example the IM window display does not fit the grid of characters well  the base character is selected. For example type `a`  or click `a` the menu of characters to choose from doesn't always fit well.

File: windows/src/developer/samples/imsample/IMSample.txt (hunk 1 above)
Comment (rc-swag) [reply windows/src/developer/samples/imsample/IMSample.txt:5]: Yes.

File: windows/src/developer/samples/imsample/imsample.cpp (hunk 4 above)
Comment (rc-swag) [reply windows/src/developer/samples/imsample/imsample.cpp:529]: Yes I can't remember for sure this one was hit or not. Basically, I started out just putting checks on the ones that were crashing but it was slow cycle. add null check; build; fun; crash; view crush dump; repeat.  So then I just went through and added them everywhere.

**Comment by rc-swag** [comment]
# User Testing
This is the keyboard to be tested.  [imsample.zip](https://github.com/keymanapp/keyman/files/7982994/imsample.zip)

I have copied the tests from #5936 and made a few modifications.  
The overall test we are aiming for is that in doing the tests below the keyboard does not cause a keyman or app crash.

## SUITE_IMSAMPLE_KEYBOARD_INLINE_MENU: 
This keyboard uses the letters `aeom` to allow IMX input. This keyboard is a bit flakey being a demo and not polished.
It also takes some setting up the following registry keys need to be added and set once installed.
The location of the registry keys are at.
`Computer\HKEY_CURRENT_USER\SOFTWARE\Keyman\Keyman Engine\Active Keyboards\imsample`
and are of type REG_DWORD
The keys are `ShowIMWindow` and `ShowIMWindowAlways` they both need to be set to 0.

In this mode IM window is not shown rather the app text is updated with a menu when either `aeom` is pressed.
For example when type `he`  the text will become `h[1ɛ 2ɜ 3ə 4e 5ɘ]` then you type the number `3` the text displayed should now be `hə`

* **TEST_IMSAMPLE_INPUT:**

- Install Imsample keyboard attached to this PR.
- Select the keyboard its name is `IMTest`
- Type some keys that are not the special keys  e.g. `t` `r` `y`
- - Type the `space` key then `a` then type 1 to get the 1st option 
- Expected Result: `try æ`
- continue to the next test

* **TEST_IMSAMPLE_INPUT_CONT:**
... contd from above
- Type 'm' and type  3 to get `###` option
- Expected Result: `try æ###`

* **TEST_IMSAMPLE_BACKSPACE:**
Using Notepad or Libreoffice 
- Type `f`
- Type `e`
- The screen should have f[1ɛ 2ɜ 3ə 4e 5ɘ]
- Now press Backspace. the whole option menu including the square brackets `[` `]` should be deleted.
- Expected result `f`


## SUITE_IMSAMPLE_KEYBOARD_IM_WINDOW: 
This keyboard uses the letters `aeom` to allow IMX input.  This time a IM window should display.
It also takes some setting up the following registry keys need to be added and set once installed.
The location of the registry keys are at.
`Computer\HKEY_CURRENT_USER\SOFTWARE\Keyman\Keyman Engine\Active Keyboards\imsample`
and are of type REG_DWORD
The keys are `ShowIMWindow` and `ShowIMWindowAlways` they both need to be set to 1.


* **TEST_IMSAMPLE_INPUT_IMW:**

- Install Imsample keyboard attached to this PR (if not done already)
- Select the keyboard its name is `IMTest`
- Type some keys that are not the special keys  e.g. `t` `r` `y` 
- Type the `space` key then `a` then click the 1st option 
- Expected Result: `try æ`
- continue to the next test

* **TEST_IMSAMPLE_INPUT_IMW_CONT:**
... contd from above
- Type 'm' and select the 3rd `###` option
- Expected Result: `try æ###`

File: windows/src/developer/samples/imsample/IMSample.txt (hunk 1 above)
Comment (mcdurdin) [reply windows/src/developer/samples/imsample/IMSample.txt:5]: :+1:

**Review by mcdurdin — APPROVED** [review]
LGTM

**Comment by MakaraSok** [comment]
##SUITE_IMSAMPLE_KEYBOARD_INLINE_MENU:

- **TEST_IMSAMPLE_INPUT: PASSED** The expected result is indeed `try æ`.
- **TEST_IMSAMPLE_INPUT_CONT: PASSED** The expected result is indeed `try æ###`.
- **TEST_IMSAMPLE_BACKSPACE: PASSED** The expected result is indeed `f`.

**Comment by MakaraSok** [comment]
## SUITE_IMSAMPLE_KEYBOARD_IM_WINDOW

- **TEST_IMSAMPLE_INPUT_IMW: FAILED** Pressing 1 doesn't produce the expected output.


- **TEST_IMSAMPLE_INPUT_IMW_CONT: FAILED** Pressing 1 doesn't produce the expected output.

**Comment by rc-swag** [comment]
> * **TEST_IMSAMPLE_INPUT_IMW_CONT: FAILED** Pressing 1 doesn't produce the expected output.
> 
@MakaraSok 
When you say "pressing 1" are you clicking it or typing the number 1 on the keyboard. Because for the IM window typing the numbers will not work you need to select the option in the IM window by clicking on it.

**Comment by MakaraSok** [comment]
> > * **TEST_IMSAMPLE_INPUT_IMW_CONT: FAILED** Pressing 1 doesn't produce the expected output.
> 
> @MakaraSok When you say "pressing 1" are you clicking it or typing the number 1 on the keyboard. Because for the IM window typing the numbers will not work you need to select the option in the IM window by clicking on it.
> 
> @keymanapp-test-bot retest SUITE_IMSAMPLE_KEYBOARD_IM_WINDOW all

There was no instruction on how to select, hence pressing the number was used to select from the option list.

**Comment by bharanidharanj** [comment]
@rc-swag  Tried to test SUITE_IMSAMPLE_KEYBOARD_IM_WINDOW part and here is my observations:
1.Installed Keyman 15.0.189-alpha-test-6187 and the Keyman Developer as it attached in the Test Artifacts link.  
2. Downloaded and installed imsample keyboard. 
3. I opened Reg Editor. I did not find REG_DWORD in the following location
Computer\HKEY_CURRENT_USER\SOFTWARE\Keyman\Keyman Engine\Active Keyboards\imsample 

So, I got stuck on the problem. 

I have attached the Screenshot for reference. 

  

However, I tried to follow the steps as it mentioned in the TEST_IMSAMPLE_INPUT_IMW. When I typed some keys like t r y, nothing will show in the Notepad application. 

I need your help to test this issue. Or Did I missed out anything?

**Comment by MakaraSok** [comment]
@bharanidharanj You are supposed to add the two keys to the location in RegEditor.

> It also takes some setting up the following registry keys need to be added and set once installed.
The location of the registry keys are at.
Computer\HKEY_CURRENT_USER\SOFTWARE\Keyman\Keyman Engine\Active Keyboards\imsample
and are of type REG_DWORD
The keys are ShowIMWindow and ShowIMWindowAlways they both need to be set to 1.

@rc-swag I've tried this a few times and it seems like the IM is inactive. Pressing any of these `aeom` won't bring up anything.

**Comment by bharanidharanj** [comment]
> @bharanidharanj You are supposed to add the two keys to the location in RegEditor.
> 
> > It also takes some setting up the following registry keys need to be added and set once installed.
> > The location of the registry keys are at.
> > Computer\HKEY_CURRENT_USER\SOFTWARE\Keyman\Keyman Engine\Active Keyboards\imsample
> > and are of type REG_DWORD
> > The keys are ShowIMWindow and ShowIMWindowAlways they both need to be set to 1.
> 
> @rc-swag I've tried this a few times and it seems like the IM is inactive. Pressing any of these `aeom` won't bring up anything.

@MakaraSok Ross and I tried to add the two keys in the above mentioned location in RegEditor. After adding the two keys, we noticed that the IM is still inactive. Typing any letter on the the Notepad does not show up anything. Anyhow, Ross said that he will look in to this issue and will update it.

**Comment by bharanidharanj** [comment]
## SUITE_IMSAMPLE_KEYBOARD_IM_WINDOW:

 * **TEST_IMSAMPLE_INPUT_IMW (FAILED):** As per Ross's suggestion I have created two Keys ShowIMWindow and ShowIMWindowAlways set the values into 0. But still the the IM Sample Keyboard is not working.

 


 * **TEST_IMSAMPLE_INPUT_IMW_CONT (FAILED):** As per Ross's suggestion I have created two Keys ShowIMWindow and ShowIMWindowAlways set the values into 0. But still the the IM Sample Keyboard is not working.

**Comment by rc-swag** [comment]
@bharanidharanj  and @MakaraSok 
I have made some changes to this keyboard so it creates the registry setting the first time the keyboard is used.
You will need to uninstall the old `IMTest` Keyboard. Restart Windows. Then install the keyboard attached to this PR in the Test Instructions. [Test specification and instructions](https://github.com/keymanapp/keyman/pull/6187#issuecomment-1027493084)

@keymanapp-test-bot retest SUITE_IMSAMPLE_KEYBOARD_IM_WINDOW

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.207-beta](https://keyman.com/downloads/releases/beta/15.0.207)

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.209-beta](https://keyman.com/downloads/releases/beta/15.0.209)

**Comment by bharanidharanj** [comment]
> @bharanidharanj and @MakaraSok I have made some changes to this keyboard so it creates the registry setting the first time the keyboard is used. You will need to uninstall the old `IMTest` Keyboard. Restart Windows. Then install the keyboard attached to this PR in the Test Instructions. [Test specification and instructions](https://github.com/keymanapp/keyman/pull/6187#issuecomment-1027493084)
> 
> @keymanapp-test-bot retest SUITE_IMSAMPLE_KEYBOARD_IM_WINDOW all

@rc-swag Okay, Thanks.

**Comment by bharanidharanj** [comment]
## SUITE_IMSAMPLE_KEYBOARD_IM_WINDOW: 

* **TEST_IMSAMPLE_INPUT_IMW (PASSED):** After setting the values to 0, I got the expected output. 
* **TEST_IMSAMPLE_INPUT_IMW_CONT (PASSED):** After setting the values to 1, I got the expected output.
